### DISCUSSION TITLE
How do i schedule the job now as the hitApi function will need email as a parameter in order to run

### DISCUSSION THREAD

**Discussion by xts-bit** [taskforcesh/bullmq#2114] [Q&A]
How do i schedule the job now as the hitApi function will need email as a parameter in order to run,
But if the email is coming from the body then How will things work?

I only want to schedule the job when the user hits the API 
```
router.post('/delete-user', (req, res) => {
    const {email} = req.body;

    // How do i schedule the job now as the hitApi function will need email as a parameter in order to run,
    // But if the email is coming from the body then How will things work?
})
const queue = new Queue('myQueue', {
    connection: {
        host: 'myhost.com',
        port: 366,
        password: 'pass'
    }
});

async function hitApi(email) {
    console.log("Working")
}
queue.add("myJob", { foo: "bar" }, { delay: 1000 });

const worker = new Worker('myQueue', async (job) => {
    await hitApi(email);
}, {
    connection: {
        host: 'myhost.com',
        port: 366,
        password: 'pass'
    },
    delay: 5000
});

worker.on('completed', (job) => {
    console.log(`Job completed`);
});

worker.on('failed', (job, err) => {
    console.error(`Job failed for URL`, err);
})
```

**Comment by manast** [comment]
It looks like you want to schedule a job to delete a user when a specific API endpoint is hit, and you're using some kind of queueing mechanism to accomplish this.

You'll need to modify the router.post handler to add the job to the queue with the required email parameter. Inside the worker, you'll need to extract the email parameter from the job data. Here's an example of how you can achieve this:

```ts
// Define the endpoint that clients will POST to
router.post('/delete-user', (req, res) => {
    const {email} = req.body;

    // Schedule the job to delete the user by adding it to the queue
    queue.add("myJob", { email }, { delay: 1000 });
    res.send("Scheduled the job to delete user with email: " + email);
});

// Define the worker that will process the job
const worker = new Worker('myQueue', async (job) => {
    // Extract the email from the job data
    const email = job.data.email;
    await hitApi(email);
}, {
    connection: {
        host: 'myhost.com',
        port: 366,
        password: 'pass'
    },
    delay: 5000
});

// Define other aspects of your code as you had before...
```

This code snippet schedules a job with the user's email whenever the /delete-user endpoint is hit. Inside the worker function, the email is extracted from the job data, and the hitApi function is called with that email.

**Reply by xts-bit** [reply]
@manast Hmm. This can be a solution, But will my code will work when the server is restarted? like as I push new code to the server it restarts the server so will this cause any problem? in my case, the delay is around 1 day. 

 As `queue.add("myJob", { email }, { delay: 1000 });` is only called inside the API which will hit only one time.

**Reply by manast** [reply]
As long as Redis is properly configured it will work after restarts. https://docs.bullmq.io/guide/going-to-production

**Reply by xts-bit** [reply]
@manast Can you please tell me what exactly to do? I'm using redis labs free database will things work?

**Reply by xts-bit** [reply]
@manast Any idea about this what to do? is it possible with redis labs?

**Reply by xts-bit** [reply]
@manast I configured maxRetriesPerRequest, enableReadyCheck, and enableOfflineQueue, Is there anything else missing that I need to configure to make things work after restart?

**Reply by roggervalf** [reply]
hey @xts-bit, I think @manast meant following https://docs.bullmq.io/guide/going-to-production and also considering that your jobs are saved in redis and you are only restarting your server (not flushing your redis db), things should work as expected. Your delayed jobs may be saved in redis and after restarts may still there until one worker pick them when delays are passed

**Reply by xts-bit** [reply]
@roggervalf Redis's free plan doesn't offer Persistence any other way please help.

**Reply by manast** [reply]
We offer a certain level of help for free, but if you misuse this gift and annoy us, we will need to take some action to make it stop. You have got a lot of information already. If you cannot afford a paid Redis instance, then most likely this library is not suitable for your needs.
